### ISSUE TITLE
[@bs] at application side treats `f ()` as 0-arity application

### ISSUE THREAD

**Issue by andreypopp** [melange-re/melange#674]
Consider this [gist](https://melange.re/unstable/playground/?language=OCaml&code=bGV0IHNsZWVwIG1zID0KICBKcy5Qcm9taXNlLm1ha2UgQEAgZnVuIH5yZXNvbHZlIH5yZWplY3Q6XyAtPgogIGlnbm9yZSAoSnMuR2xvYmFsLnNldEludGVydmFsIChmdW4gKCkgLT4gKHJlc29sdmUgKCkgW0Bic10pKSBtcyk%3D&live=off):
```
let sleep ms =
  Js.Promise.make @@ fun ~resolve ~reject:_ ->
  ignore (Js.Global.setInterval (fun () -> (resolve () [@bs])) ms)
```
which errors with
```
Line 3, 44: Error This expression has type ('a -> unit)   
  Js__Js_internal.Fn.arity1 but an expression was expected of type 'b  
  Js__Js_internal.Fn.arity0
```

The issue is that Melange treats `f ()[@bs]` syntax as 0-arity application.
If we do
```
let unit = () in f unit [@bs]
```
then it compiles.

I think 0-arity applications should have another syntax.

**Comment by anmonteiro** [comment]
I took a look at this. Here are some thoughts:

- `f () [@u]` (previoulsy `[@bs]`) is a syntactical transformation
  - if the only argument is `()`, melange considers it a 0-arity application
- We're making a few breaking changes for Melange 2.0. If desirable and agreed upon, changing the 0-arity application syntax is possible

Here are some suggestions for a new 0-arity application syntax:

1. introduce a new attribute:

```ocaml
let () = f () [@u0] (* or any other attribute different than [@u] *)
```

2. introduce a new module:

```ocaml
let () = Uncurried.apply0 f
```

3. introduce a new special operator:

```ocaml
let () = f $$u
```

4. introduce a new let operator:

```ocaml
let>$ () = f ()
```

5. introduce a new extension point:

```ocaml
let%u0 () = f ()
```

**Comment by jchavarri** [comment]
Sorry for the probably stupid questions, but I never understood this:

- Why is it necessary to mark functions with arity 1 with `@u`?
- What does "arity 0" mean? If a function has no arguments, then it's not a function, right?

**Comment by anmonteiro** [comment]
- `f () [@u]` becomes `Js.Internal.run f`, compiles to `f()` in JS
- `f arg1 .. argN [@u]` becomes `Js.Internal.Fn.arityN`, compiles to `f(arg1, .., argN)`


so 0-arity is different than n-arity for uncurried because:
- in OCaml it still has 1 argument
- we want to compile to `f()` and not `f(undefined)`
- this needs to be decided at PPX level

**Comment by andreypopp** [comment]
Is it possible to use `f [@u0]`? 

Otherwise my vote is for option 1.

**Comment by anmonteiro** [comment]
> I am also surprised this didn't come up earlier?

It did: https://github.com/rescript-lang/rescript-compiler/issues/3429

**Comment by anmonteiro** [comment]
> 

It's technically possible, but that doesn't look like function application anymore. Something to ponder.

**Comment by anmonteiro** [comment]
This seems to come up especially for `Promise.make`, which is defined as:

```ocaml
external make :
  ((resolve:(('a -> unit)[@u]) -> reject:((exn -> unit)[@u]) -> unit)
  [@mel.uncurry]) ->
  'a t = "Promise"
[@@mel.new]
```

However, both `resolve` and `reject` only take 1 argument each, and they probably don't need to be curried at all.

So we could fix the immediate case with

```ocaml
external make :
  ((resolve:('a -> unit) -> reject:(exn -> unit) -> unit)[@mel.uncurry]) -> 'a t
  = "Promise"
[@@mel.new]
```

This would be a breaking change, but so would changing the 0-arity application syntax.

**Comment by anmonteiro** [comment]
OK, I suppose the issue occurs when you pass `resolve` as an argument. Consider the following program:

```ocaml
type 'a t

external make :
  ((resolve:('a -> unit) -> reject:(exn -> unit) -> unit)[@mel.uncurry]) -> 'a t
  = "Promise"
[@@mel.new]

external makeU :
  ((resolve:(('a -> unit)[@u]) -> reject:((exn -> unit)[@u]) -> unit)
  [@mel.uncurry]) ->
  'a t = "Promise"
[@@mel.new]

let r resolve =
  Js.log "resolve normal";
  resolve "normal"

let rU resolve =
  Js.log "resolve uncurried";
  resolve "uncurried" [@u]

let x = make (fun ~resolve ~reject -> r resolve)
let xU = makeU (fun ~resolve ~reject -> rU resolve)
```

and the generated JS:

```js
$ melc -ppx melppx x.ml

melc -ppx melppx x.ml
// Generated by Melange
'use strict';

var Curry = require("melange.js/curry.js");

function r(resolve) {
  console.log("resolve normal");
  return Curry._1(resolve, "normal");
}

function rU(resolve) {
  console.log("resolve uncurried");
  return resolve("uncurried");
}

var x = new Promise((function (resolve, param) {
        r(resolve);
      }));

var xU = new Promise((function (resolve, param) {
        rU(resolve);
      }));

exports.r = r;
exports.rU = rU;
exports.x = x;
exports.xU = xU;
/* x Not a pure module */
```

**Comment by jchavarri** [comment]
Hm, and it seems that `mel.uncurry` only works with the "outer" type of the external, but not to annotate "internal" types like the labelled `resolve`: [example](https://melange.re/unstable/playground/?language=OCaml&code=dHlwZSAnYSB0CgpleHRlcm5hbCBtYWtlIDoKICAoKHJlc29sdmU6KCgnYSAtPiB1bml0KVtAYnMudW5jdXJyeV0pIC0%2BIHJlamVjdDooZXhuIC0%2BIHVuaXQpIC0%2BIHVuaXQpW0Bicy51bmN1cnJ5XSkgLT4gJ2EgdAogID0gIlByb21pc2UiCltAQGJzLm5ld10KCmxldCByIHJlc29sdmUgPQogIEpzLmxvZyAicmVzb2x2ZSBub3JtYWwiOwogIHJlc29sdmUgIm5vcm1hbCIKCmxldCB4ID0gbWFrZSAoZnVuIH5yZXNvbHZlIH5yZWplY3QgLT4gciByZXNvbHZlKQo%3D&live=off).

**Comment by anmonteiro** [comment]
I propose we postpone this for v3 to give us some more room for brainstorming.

In the meantime, we should probably document this in the website as an edge case.

**Comment by anmonteiro** [comment]
#949 is progressing nicely, though a bit more invasive than I'd like. I didn't expect this would touch so much code.

There are currently 2 open questions:

- what to do about `[@mel.meth]`?
  - it currently uses the same detection as `[@u]` and a single unit argument. A few options:
    1. `[mel.meth0]` doesn't have the same ring to it as `[@u0]`
    2. `[mel.meth]` and `[@u0]` together -- this becomes a bit harder to explain since `[@mel.meth]` is assumed to be uncurried and doesn't need `[@u]` otherwise
- this is quite the breaking change for code using `[@u]` and `unit` arguments. If we merge #949, every function that was inferred to be 0-arity will now become 1-arity with an `undefined` argument (because that's what `unit` compiles to). I think this one can be easily detected and add an upgrade path:
  1. add a transitory alert when detecting `unit -> 'a` and `[@u]`, warning that users probably want `[@u0]` instead

**Comment by jchavarri** [comment]
> * `[mel.meth0]` doesn't have the same ring to it as `[@u0]`

What about `mel.meth_no_params`? I think we should stay consistent, as you mention in the PR.

> 1. add a transitory alert when detecting `unit -> 'a` and `[@u]`, warning that users probably want `[@u0]` instead

This sounds great.

**Comment by anmonteiro** [comment]
> 1. add a transitory alert when detecting `unit -> 'a` and `[@u]`, warning that users probably want `[@u0]` instead

I had forgotten to add a test showing what the alert looks like. Added a commit to #949: https://github.com/melange-re/melange/pull/949/commits/66a247ab9a6143c3676c8b9de58bafdf73b96d96

**Comment by anmonteiro** [comment]
> What about `mel.meth_no_params`?

probably worth it to be consistent indeed. I don't like `mel.meth_no_params` either.

Another option I thought about was `@mel.meth_void`, but I'd probably still prefer `@mel.meth0` in that case.

**Comment by jchavarri** [comment]
I also prefer `meth0` to `meth_void` fwiw.

**Comment by anmonteiro** [comment]
In our meeting today we decided to:

- drop the work for `@mel.meth0`
- drop `@mel.meth` entirely from the melange ffi
  - perhaps deprecate first, then remove for v4

**Comment by anmonteiro** [comment]
I had an idea that could unblock the work on 0-arity application -- it's a breaking change, so we'd likely need a new attribute or to introduce it in a breaking release, but essentially:

- Uncurried function with `[@u]` have **at least** arity of 1
- Zero-arity function must be annotated with `[@u 0]`

I think this would generally work, and we could likely enforce that a `[@u 0]` can only be applied with `()`.

What do you think?
